src/mesh/voxel_cloud.rs

```diff
@@ -448,6 +448,165 @@ impl ScalarField {
+    /// risk of generating a non-manifold mesh.
+    pub fn to_marching_cubes<U>(&self, volume_value_range: &U) -> Option<Mesh>
+    where
+        U: RangeBounds<f32>,
+    {
+        if self.block_dimensions.x == 0
+            || self.block_dimensions.y == 0
+            || self.block_dimensions.z == 0
+            || !self.contains_voxels_within_range(volume_value_range)
+        {
+            return None;
+        }
+
+        let marching_cubes_lookup_table = marching_cubes_lookup_table();
+
+        // A collection of mesh patches defining an outer envelope of volumes
+        // stored in the scalar field
+        let mut marching_cubes: Vec<Mesh> = Vec::new();
+
+        let block_end = self.block_end();
+
+        // Iterate through the scalar field. Consider 8 neighboring voxels to be
+        // corners of the test cube. Start iterating at position [-1, -1, -1]
+        // relative to the `block_start` and finish at [1, 1, 1] relative to the
+        // `block_end` to catch the block boundaries.
+        for z in (self.block_start.z - 1)..=block_end.z {
+            for y in (self.block_start.y - 1)..=block_end.y {
+                for x in (self.block_start.x - 1)..=block_end.x {
+                    //                 v4_______e4_____________v5
+                    //                  /|                    /|
+                    //                 / |                   / |
+                    //              e7/  |                e5/  |
+                    //               /___|______e6_________/   |
+                    //            v7|    |                 |v6 |e9
+                    //              |    |                 |   |
+                    //              |    |e8               |e10|
+                    //           e11|    |                 |   |
+                    //              |    |_________________|___|
+                    //              |   / v0    e0         |   /v1
+                    //              |  /                   |  /
+                    //              | /e3                  | /e1
+                    //              |/_____________________|/
+                    //              v3         e2          v2
+                    //
+                    // the current voxel center is `v3`, the other vertices are centers of
+                    // the neighboring voxels
+                    let corners_absolute_coordinates = [
+                        Point3::new(x, y + 1, z),
+                        Point3::new(x + 1, y + 1, z),
+                        Point3::new(x + 1, y, z),
+                        Point3::new(x, y, z),
+                        Point3::new(x, y + 1, z + 1),
+                        Point3::new(x + 1, y + 1, z + 1),
+                        Point3::new(x + 1, y, z + 1),
+                        Point3::new(x, y, z + 1),
+                    ];
+                    let corners_cartesian_coordinates: Vec<_> = corners_absolute_coordinates
```

Comment:
No need to heap-allocate. Either collect to a `ArrayVec<[Point<f32>; 8]>`, or perhaps just explicitly write out the array.

I prefer the latter, because the only reason we currently use the arrayvec crate is by-value `.into_iter()` for arrays (intead of by-ref), but rust is considering to fix this with a breaking change (https://github.com/rust-lang/rust/pull/65819), meaning we could potentially drop arrayvec as a dependency later.